toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -660,3 +722,78 @@ func getPartUUID(device string) (uuid string, err error) {
 	uuid = strings.TrimSpace(stdout)
 	return
 }
+
+func checkSectorSizeFile(sectorFile string) (sectorSize uint64, err error) {
+	if exists, ferr := file.PathExists(sectorFile); ferr != nil {
+		logger.Log.Errorf("Error accessing sector size file %s", sectorFile)
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("could not find the hw sector size file %s to obtain the sector size of the system", sectorFile)
+		return
+	}
+
+	fileContent, err := file.ReadLines(sectorFile)
+	if err != nil {
+		logger.Log.Errorf("Failed to read from %s: %s", sectorFile, err)
+		return
+	}
+
+	// sector file should only have one line, return error if not
+	if len(fileContent) != 1 {
+		err = fmt.Errorf("%s has more than one line", sectorFile)
+		return
+	}
+
+	sectorSize, err = strconv.ParseUint(fileContent[0], 10, 64)
+	return
+}
+
+func getSectorSize(diskDevPath string) (logicalSectorSize, physicalSectorSize uint64, err error) {
+	// Grab the specific disk name from /dev/xxx
+	diskName := diskDevPath[5:len(diskDevPath)]
+
+	hw_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/hw_sector_size", diskName)
+	physical_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/physical_block_size", diskName)
+
+	logicalSectorSize, err = checkSectorSizeFile(hw_sector_size_file)
+	if err != nil {
+		return
+	}
+
+	physicalSectorSize, err = checkSectorSizeFile(physical_sector_size_file)
+	return
+}
+
+func alignSectorAddress(sectorAddr, physicalSectorSize uint64) (alignedSector uint64) {
+	if sectorAddr%physicalSectorSize == 0 || sectorAddr < physicalSectorSize {
+		alignedSector = sectorAddr
+	} else {
+		alignedSector = (sectorAddr/physicalSectorSize + 1) * physicalSectorSize
+	}
+
+	return
+}
+
+func obtainPartitionDetail(partitionIndex int, hasExtendedPartition bool) (partType string, partitionNumber int) {
+	const (
+		primaryPartition  = "primary"
+		extendedPartition = "extended"
+		logicalPartition  = "logical"
+	)
+
+	if hasExtendedPartition && partitionIndex >= (maxPrimaryPartitionsForMBR-1) {
```

Comment:
Bit of code smell here. Is it always going to be the case where we can only define 1 extended partition for an MBR disk? Why use +2 for logical partitions?

Reply:
Did a little research, which I probably should have done before my comment here. So with MBR, one can have 4 primary partitions or 3 primary partitions + extended partition with N number of logical partitions inside.

I still don't quite understand the +1 and +2 logic here.

Reply:
Yes, only 1 extended partition is allowed in one disk. 

+2 for the logical partition is due to the insertion of the extended partition here. The extended partition needs to be created before logical partitions can be allocated. E.g. partition 4 now will be occupied by the extended partition and thus the 4th partition specified by the users will now occupy partition number 5, which +2 based on its index in the partition list.

Reply:
Makes sense but let's use well-named constants instead of "magic numbers" in that case. This will help with understanding that logic to other devs. If you think the names would still not be clear enough, we can add comments on top of that.

Reply:
Added constants and comments for better explanation.